include/rmm/mr/device/limiting_resource_adaptor.hpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/detail/error.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+
+namespace rmm {
+namespace mr {
+/**
+ * @brief Resource that uses `Upstream` to allocate memory and limits the total
+ * allocations possible.
+ *
+ * An instance of this resource can be constructed with an existing, upstream
+ * resource in order to satisfy allocation requests, but any existing allocations
+ * will be untracked.
+ *
+ * @tparam Upstream Type of the upstream resource used for
+ * allocation/deallocation.
+ */
+template <typename Upstream>
+class limiting_resource_adaptor final : public device_memory_resource {
+ public:
+  /**
+   * @brief Construct a new limiting resource adaptor using `upstream` to satisfy
+   * allocation requests and limiting the total allocation amount possible.
+   *
+   * @throws `rmm::logic_error` if `upstream == nullptr`
+   *
+   * @param upstream The resource used for allocating/deallocating device memory
+   * @param allocation_limit Maximum memory allowed for this allocator.
+   */
+  limiting_resource_adaptor(Upstream* upstream,
+                            std::size_t allocation_limit_ = std::numeric_limits<std::size_t>::max(),
+                            std::size_t allocation_alignment_ = 256)
```

Comment:
It certainly does, but we need to know the alignment of the upstream and they have different alignments. It would be nice to have a common way to query the upstream alignment, but not all upstreams have the ability to ask. I didn't want to jump on an opinionated approach here, so I was going to make it an option and open an issue for discussion. There are other comments around related to alignment such as `// TODO use rmm-level def of this.` in `pool_memory_resource.hpp` that make me think a discussion is warranted.